### ISSUE TITLE
macOS/Metal: low performance

### ISSUE THREAD

**Issue by divVerent** [hajimehoshi/ebiten#2189]
Similar to #2188, I did the same tests on a Macbook Pro (14-inch, 2021) and got:

```
cp benchmark.dem /tmp/
time open -W --env EBITEN_GRAPHICS_LIBRARY=opengl ./AAAAXY.app --args -load_config=false -vsync=false -demo_play=/tmp/benchmark.dem -demo_timedemo
5.091 total
time open -W --env EBITEN_GRAPHICS_LIBRARY=metal ./AAAAXY.app --args -load_config=false -vsync=false -demo_play=/tmp/benchmark.dem -demo_timedemo
16.376 total
```

FPS can be shown via `open --env EBITEN_GRAPHICS_LIBRARY=opengl ./AAAAXY.app --args -load_config=false -vsync=false -show_fps` and are 670fps for OpenGL and 110fps for Metal. Note that 110fps are quite much playable, so this isn't really a big issue on this particular hardware. Just thought it to be noteworthy as there may be issues on lower-performance Macs.

**Comment by hajimehoshi** [comment]
What draw calls are executed? You can see them with `-tags=ebitendebug`.

**Comment by divVerent** [comment]
Don't have the draw call list for MacOS right here, but on OpenGL/Linux I get these, and the exact same on DirectX/Windows, which makes me think this output is OS and backend independent:

https://github.com/hajimehoshi/ebiten/issues/2188#issuecomment-1179528460

**Comment by hajimehoshi** [comment]
Could you try the latest 2.3 version `e488486d97107ae631aea45c8bdb2e73eb7c0f17`?

**Comment by hajimehoshi** [comment]
I tested this with vsync=off. Actually Metal was 2x slower than OpenGL. I tried to make `(g *Graphics) DrawTriangles` empty and interestingly Metal was 2x faster than OpenGL in this case. This means that we have to optimize the implementation of `(g *Graphics) DrawTriangles` for Metal, like reducing allocating float32 slices.

**Comment by hajimehoshi** [comment]
Hmm, reducing slices doesn't solve this issue.

On the fullscreen, the FPS is almost the same (like 110-120) with Metal and OpenGL on my MacBook Pro. So, probably this is a Metal-specific issue with vsync=off, but I'm not sure.

**Comment by divVerent** [comment]
To be honest, I cannot rule out this merely being a vsync=off problem, as I
never drop below 60fps on my macbook. I don't have an older device to try
rule that out.

This issue is also IMHO way lower priority than the Windows one, as macOS
users tend to not have low-end hardware.

**Comment by hajimehoshi** [comment]
I agree this is low priority. I'll leave this open without a milestone

**Comment by hajimehoshi** [comment]
I'll add a compiler flag to optimize shaders: https://developer.apple.com/documentation/metal/mtlcompileoptions?language=objc

**Comment by hajimehoshi** [comment]
Hm, the optimization level https://developer.apple.com/documentation/metal/mtlcompileoptions/3929909-optimizationlevel?language=objc is still beta.

**Comment by tinne26** [comment]
I don't have any macOS device to test on, but as a small suggestion: what about using an example like [sprites](https://github.com/hajimehoshi/ebiten/tree/main/examples/sprites) to eventually overload the system and at least determine whether it's vsync-off related or not? If it's vsync-off related we can all agree it's not a big deal, but otherwise it may deserve more consideration.

**Comment by hajimehoshi** [comment]
Removed the milestone as there are no clear action items.
